src/App.tsx

```diff
@@ -1,14 +1,72 @@
 /* eslint-disable max-len */
-import React from 'react';
+import { useEffect, useMemo, useState } from 'react';
 import 'bulma/css/bulma.css';
 import '@fortawesome/fontawesome-free/css/all.css';
 
 import { TodoList } from './components/TodoList';
 import { TodoFilter } from './components/TodoFilter';
 import { TodoModal } from './components/TodoModal';
 import { Loader } from './components/Loader';
+import { getTodos } from './api';
+import { Todo } from './types/Todo';
 
 export const App: React.FC = () => {
+  const [isLoading, setIsLoading] = useState(true);
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [selectedTodo, setSelectedTodo] = useState<Todo | null>(null);
+  const [todoSelector, setTodoSelector] = useState('all');
+  const [todoFilter, setTodoFilter] = useState('');
+
+  useEffect(() => {
+    getTodos().then((data) => {
+      setTodos(data);
+      setIsLoading(false);
+    });
+  }, []);
+
+  const getVisibleTodos = () => {
+    return todos.filter((todo) => {
+      const todoTitleIncludesFilter = todo.title
+        .toLowerCase()
+        .includes(todoFilter.toLowerCase());
+
+      switch (todoSelector) {
+        case 'active':
+          return !todo.completed && todoTitleIncludesFilter;
+        case 'completed':
+          return todo.completed && todoTitleIncludesFilter;
+        default:
+          return todoTitleIncludesFilter;
+      }
+    });
+  };
+
+  const visibleTodos = useMemo(getVisibleTodos, [
+    todos,
+    todoSelector,
+    todoFilter,
+  ]);
```

Comment:
This wont work. We need useMemo to avoid unnesessary loop iteration. Also if selected all and query is empty - no need to map all todos. You should do like this.
```suggestion
  const visibleTodos = useMemo(() => {
    const needsToFilter = todoSelector === 'active' || todoSelector === 'completed';
  
    if (!needsToFilter || !todoFilter) {
      return todos;
    }
    
    return todos.filter((todo) => {
      const todoTitleIncludesFilter = todo.title
        .toLowerCase()
        .includes(todoFilter.toLowerCase());
        
        switch (todoSelector) {
        case 'active':
          return !todo.completed && todoTitleIncludesFilter;
        case 'completed':
          return todo.completed && todoTitleIncludesFilter;
        default:
          return todoTitleIncludesFilter;
      }     
    });
 
  }, [todos, todoSelector, todoFilter,]);

```

Reply:
> This wont work. We need useMemo to avoid unnesessary loop iteration. Also if selected all and query is empty - no need to map all todos. You should do like this.

Requested changes are added